witness/golang/internal/persistence/inmemory/inmemory.go

```diff
@@ -0,0 +1,133 @@
+// Copyright 2022 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package inmemory provides a persistence implementation that lives only in memory.
+package inmemory
+
+import (
+	"fmt"
+	"reflect"
+	"sync"
+
+	"github.com/google/trillian-examples/witness/golang/internal/persistence"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// NewInMemoryPersistence returns a persistence object that lives only in memory.
+func NewInMemoryPersistence() persistence.LogStatePersistence {
+	return &inMemoryPersistence{
+		checkpoints: make(map[string]checkpointState),
+	}
+}
+
+type checkpointState struct {
+	rawChkpt     []byte
+	compactRange []byte
+}
+
+type inMemoryPersistence struct {
+	mu          sync.RWMutex
+	checkpoints map[string]checkpointState
+}
+
+func (p *inMemoryPersistence) Init() error {
+	return nil
+}
+
+func (p *inMemoryPersistence) Logs() ([]string, error) {
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+	res := make([]string, 0, len(p.checkpoints))
+	for k := range p.checkpoints {
+		res = append(res, k)
+	}
+	return res, nil
+}
+
+func (p *inMemoryPersistence) ReadOps(logID string) (persistence.LogStateReadOps, error) {
+	p.mu.RLock()
+	defer p.mu.RUnlock()
```

Comment:
Would this be more concise/closer fit the semantics of the interface if the `[R]Unlock` calls were made in the `readWriter.Close` method?

Reply:
This would be:
 a) more dangerous
 b) not in the contract of the reader (it doesn't have a close method)
 c) slower for performance

This method reads the map once and puts the value into a local field. I think that matches the transactional semantics pretty well?